assembly/src/ast/module.rs

```diff
@@ -1,321 +1,718 @@
-        self.import_info = import_info;
+    /// When a source file is given, diagnostics will contain source code snippets.
+    pub fn with_source_file(mut self, source_file: Option<Arc<SourceFile>>) -> Self {
+        self.source_file = source_file;
         self
     }
 
-    // PARSER
-    // --------------------------------------------------------------------------------------------
-    /// Parses the provided source into a [ModuleAst].
-    ///
-    /// A module consists of internal and exported procedures but does not contain a body.
-    pub fn parse(source: &str) -> Result<Self, ParsingError> {
-        let mut tokens = TokenStream::new(source)?;
-        let mut import_info = ModuleImports::parse(&mut tokens)?;
-        let local_constants = parse_constants(&mut tokens)?;
-        let mut context = ParserContext {
-            import_info: &mut import_info,
-            local_procs: LocalProcMap::default(),
-            reexported_procs: ReExportedProcMap::default(),
-            local_constants,
-            num_proc_locals: 0,
-        };
-        context.parse_procedures(&mut tokens, true)?;
+    /// Specifies the source span in the source file in which this module was defined, that covers
+    /// the full definition of this module.
+    pub fn with_span(mut self, span: SourceSpan) -> Self {
+        self.span = span;
+        self
+    }
 
-        // make sure program body is absent and there are no more instructions.
-        if let Some(token) = tokens.read() {
-            if token.parts()[0] == Token::BEGIN {
-                return Err(ParsingError::not_a_library_module(token));
-            } else {
-                return Err(ParsingError::dangling_ops_after_module(token));
-            }
+    /// Like [Module::with_source_file], but does not require ownership of the [Module].
+    pub fn set_source_file(&mut self, source_file: Arc<SourceFile>) {
+        self.source_file = Some(source_file);
+    }
+
+    /// Sets the [LibraryPath] for this module
+    pub fn set_path(&mut self, path: LibraryPath) {
+        self.path = path;
+    }
+
+    /// Sets the [LibraryNamespace] for this module
+    pub fn set_namespace(&mut self, ns: LibraryNamespace) {
+        self.path.set_namespace(ns);
+    }
+
+    /// Sets the documentation for this module
+    pub fn set_docs(&mut self, docs: Option<Span<String>>) {
+        self.docs = docs;
+    }
+
+    /// Like [Module::with_span], but does not require ownership of the [Module].
+    pub fn set_span(&mut self, span: SourceSpan) {
+        self.span = span;
+    }
+
+    /// Defines a procedure, raising an error if the procedure is invalid, or conflicts with a
+    /// previous definition
+    pub fn define_procedure(&mut self, export: Export) -> Result<(), SemanticAnalysisError> {
+        if self.is_kernel() && matches!(export, Export::Alias(_)) {
+            return Err(SemanticAnalysisError::ReexportFromKernel {
+                span: export.span(),
+            });
+        }
```

Comment:
What is the motivation for this restriction? I don't think we currently use this - so, it shouldn't break anything. But I'm not seeing reasons for why this should be prohibited.

Reply:
It probably isn't essential, no. Originally, it was because I was treating everything exported from a kernel as `syscall`-able, but that's not the case anymore, so in theory we don't need to restrict it.

I still think there is some question of what it means to export something from the kernel that _isn't_ a syscall, so keeping it restricted for now is maybe best, but I'll leave that up to you.

Reply:
The way I think about it is that any procedure exported from the kernel module is `syscall`-able - this includes both procedures defined in the module itself as well as re-exported from other modules. I'm not sure the above restriction is actually meaningful because instead of:
```
use.foo::bar

export.bar::baz
```
We can always write:
```
use.foo::bar

export.baz
    exec.bar::baz
end
```